cupyx/scipy/signal/_peak_finding.py

```diff
@@ -1492,3 +1496,440 @@ def argrelextrema(data, comparator, axis=0, order=1, mode="clip"):
             "CuPy `take` doesn't support `mode='raise'`.")
 
     return cupy.nonzero(results)
+
+
+_reduce_peaks = cupy.RawKernel(r"""
+#include <cupy/math_constants.h>
+#define MAX_SZ 1024
+
+extern "C" __global__ void reduce_peaks(
+        const int row_idx,
+        const int gap_thresh,
+        const double* __restrict__ max_distances,
+        const long long* __restrict__ rows,
+        const long long* __restrict__ peaks_cols,
+        const long long* __restrict__ peaks_per_row,
+        const long long* __restrict__ peak_row_offsets,
+        const long long* __restrict__ ridge_gap,
+        const int n_ridges,
+        long long* peak_ridges,
+        long long* ridges_col,
+        long long* valid_ridges,
+        long long* row_ridges) {
+
+    extern __shared__ int sh_ridge_gap[MAX_SZ];
+    extern __shared__ long long ridge_lines[MAX_SZ];
+    extern __shared__ long long ridge_line_cols[MAX_SZ];
+
+    const int base_idx = blockDim.x * blockIdx.x + threadIdx.x;
+
+    const long long row = rows[row_idx];
+    const long long row_peaks = peaks_per_row[row_idx];
+    const long long row_peak_offset = peak_row_offsets[row_idx];
+    const long long idx = base_idx + row_peak_offset;
+
+    const int ridge_partitions = (n_ridges + blockDim.x - 1) / blockDim.x;
+
+    double dist = max_distances[row] + 1e-6;
+    const long long col = peaks_cols[idx];
+    long long chosen_ridge = -1;
+
+    for(int i = 0; i < ridge_partitions; i++) {
+        int ridge_sz = blockDim.x;
+        if(i == ridge_partitions - 1 && n_ridges % blockDim.x > 0) {
+            ridge_sz = n_ridges % blockDim.x;
+        }
+
+        if(threadIdx.x < ridge_sz) {
+            const int ridge_idx = blockDim.x * i + threadIdx.x;
+            ridge_lines[threadIdx.x] = valid_ridges[ridge_idx];
+            ridge_line_cols[threadIdx.x] = ridges_col[ridge_idx];
+            sh_ridge_gap[threadIdx.x] = ridge_gap[ridge_idx];
+        }
+
+        __syncthreads();
+
+        if(base_idx < row_peaks) {
+            for(int j = 0; j < ridge_sz; j++) {
+                const long long gap = sh_ridge_gap[j];
+                if(gap <= gap_thresh) {
+                    const long long other_col = ridge_line_cols[j];
+                    double diff = abs(col - other_col);
+                    if(diff <= max_distances[row]) {
+                        if(diff < dist) {
+                            chosen_ridge = ridge_lines[j];
+                            dist = diff;
+                        }
+                    }
+                }
+            }
+        }
+
+        __syncthreads();
+    }
+
+    if(base_idx < row_peaks) {
+        if(chosen_ridge != -1) {
+            peak_ridges[idx] = chosen_ridge;
+            row_ridges[base_idx] = chosen_ridge;
+        } else {
+            row_ridges[base_idx] = idx;
+        }
+    }
+}
+""", 'reduce_peaks')
+
+
+def _identify_ridge_lines(matr, max_distances, gap_thresh):
```

Comment:
After many weeks and attempts to parallelize this function, which tries to join points in a 2D plane in order to form "ridge lines" for a CWT, I've come to the conclusion that the data dependency is to high in order for it to behave exactly like SciPy.

I tried several alternatives such as:
1. Let each point find its closest one, then merge the lines (fully parallel) → [3763d6d](https://github.com/cupy/cupy/pull/7875/commits/3763d6da5e4e239c5c36b4948c02ddfb89d1a154)
2. Let each point find its closest one in a window in a single kernel (fully parallel) → https://github.com/cupy/cupy/pull/7875/commits/503acc36bde01f908e91993dcc671ddac33a2892
3. For each row, let each point find its closest line from the previous row (SciPy behaviour, data dependency between row results, dynamic size of lines). → [1f9147c](https://github.com/cupy/cupy/pull/7875/commits/1f9147c59b5ac960b571049201cd21f4040511be)

While 1 and 2 are parallel, they do not match the SciPy behaviour 100% and are invalid with respect the reference paper upon SciPy is based on, which causes some mismatches in some tests, and therefore potentially not complying with the `find_peaks_cwt` signature. 3 can parallelize the per-row matching, at the expense of having to recompute the current valid lines on each iteration, since new ones can appear dynamically, which cost more than .

I don't know if we are willing to expose a way slow API in CuPy that implements `find_peaks_cwt` or just omit it altogether in this case, as the benchmark results are not convincing enough

Reply:
I guess some algorithms just cannot be parallelized, esp within the backcompat constraints. For me, your attempts are definitely convincing that this is one of these cases!

So I guess possible options are:
- leave this function out, or
- translate the python code in scipy with numpy->cupy and leave it at that

In other parts of cupyx.scipy.signal we took the second option in the name of API coverage.

Reply:
It would be acceptable to take the second option as @ev-br suggested here. Thank you for your exploration of the alternatives! These could be used as the basis for a later improvement.